src/identity/customca.rs

```diff
@@ -0,0 +1,129 @@
+use super::CertificateProvider;
+use async_trait::async_trait;
+use kube::{
+    api::{Api, PostParams, ListParams, WatchEvent},
+    client::Client,
+};
+use tracing::{instrument, warn};
+use crate::{identity::manager::Identity, tls::SanChecker};
+use crate::identity::Error;
+use crate::tls;
+use futures::{StreamExt, TryStreamExt};
+use k8s_openapi::{api::certificates::v1::{CertificateSigningRequest, CertificateSigningRequestSpec}, ByteString};
+use std::str;
+use rand::prelude::*;
+
+#[derive(Clone)]
+pub struct CustomSigner {
+    pub signer: Option<std::string::String>,
+}
+
+impl CustomSigner {
+    pub fn new(signer: Option<std::string::String>) -> Result<CustomSigner, Error> {
+        Ok(CustomSigner { signer })
+    }
+}
+
+#[async_trait]
+impl CertificateProvider for CustomSigner {
+    #[instrument(skip_all)]
+    async fn fetch_certificate(&self, id: &Identity) -> Result<tls::Certs, Error> {
```

Comment:
Can you document what is trigger fetch_certificate()? And how is certificate being refreshed?

Reply:
If cert is already there in cache we don't need to create a CSR but if it's not there, we refresh the cache by creating a new K8s CSR. It can be triggered in 3 ways - 
1) Pod start-up - pre-load them so that first req shouldn't be slow.
2) Inbound Req
3) Outbound Req
@howardjohn please correct me if I missed anything here.

Reply:
Cache is handling that case already 
https://github.com/istio/ztunnel/blob/master/src/identity/manager.rs#L233
https://github.com/istio/ztunnel/blob/master/src/tls/boring.rs#L165